lib/common/acl.c

```diff
@@ -781,3 +790,363 @@ pcmk__update_acl_user(xmlNode *request, const char *field,
+                }
+                xmlSetNs(i_node, ns_recycle_readable);
+            } else if (ns == NS_DENIED) {
+                if (ns_recycle_denied == NULL) {
+                    ns_recycle_denied = xmlNewNs(xmlDocGetRootElement(i_node->doc),
+                                                 NS_DENIED, ACL_NS_Q_DENIED);
+                    ret |= PCMK_ACL_VERDICT_DENIED;
+                }
+                xmlSetNs(i_node, ns_recycle_denied);
+            }
+            break;
+        default:
+            break;
+        }
+    }
+
+    return ret;
+}
+
+int
+pcmk_acl_evaled_as_namespaces(enum pcmk_acl_cred_type cred_type,
+                              const char *cred, xmlDoc *cib_doc,
+                              xmlDoc **acl_evaled_doc)
+{
+    /* XXX requires prior crm_xml_init */
+
+    int ret, version;
+    xmlNode *target, *comment;
+    char comment_buf[256] = "access as evaluated for user ";
+    const char *validation;
+
+    CRM_CHECK(cred != NULL, return -1);
+    CRM_CHECK(cib_doc != NULL, return -1);
+    CRM_CHECK(acl_evaled_doc != NULL, return -1);
+    /* XXX no proper support for groups yet */
+    CRM_CHECK(cred_type == PCMK_ACL_CRED_USER, return -2);
+
+    if (!pcmk_acl_required(cred)) {
+        /* nothing to evaluate */
+        return 0;
+    }
+
+    /* XXX see the comment for this function, pacemaker-4.0 may need
+           updating respectively in the future */
+    validation = crm_element_value(xmlDocGetRootElement(cib_doc),
+                                   XML_ATTR_VALIDATION);
+    version = get_schema_version(validation);
+    if (get_schema_version(PCMK_COMPAT_ACL_2_MIN_INCL) > version
+            || (-1 != get_schema_version(PCMK_COMPAT_ACL_2_MAX_EXCL)
+                    && get_schema_version(PCMK_COMPAT_ACL_2_MAX_EXCL)
+                        <= version)) {
+        return -3;
+    }
+
+    target = copy_xml(xmlDocGetRootElement(cib_doc));
+    if (target == NULL) {
+        return -1;
+    }
+
+    pcmk__unpack_acl(target, target, cred);
+    pcmk__set_xml_flag(target, xpf_acl_enabled);
+    pcmk__apply_acl(target);
+    ret = pcmk__eval_acl_as_namespaces_2(target);  /* XXX may need "switch" */
+
+    if (ret > 0) {
+        /* avoid trivial accidental XML injection */
+        if (strpbrk(cred, "<>&") == NULL) {
+            snprintf(comment_buf + strlen(comment_buf),
+                     sizeof(comment_buf) - strlen(comment_buf), "%s", cred);
+            comment = xmlNewDocComment(target->doc, (pcmkXmlStr) comment_buf);
+            if (comment == NULL) {
+                xmlFreeNode(target);
+                return -1;
+            }
+            xmlAddPrevSibling(xmlDocGetRootElement(target->doc), comment);
+        }
+        *acl_evaled_doc = target->doc;
+    } else {
+        xmlFreeNode(target);
+    }
+    return ret;
+}
+
+/* this is used to dynamically adapt to user-modified stylesheet */
+static const char **
+parse_params(xmlDoc *doc, const char **fallback)
```

Comment:
Figure out what this function actually does; I'm not sure we need it since we're not porting the user configuration stuff

Reply:
We don't need it; it's only used to parse the contents of a user-provided stylesheet.